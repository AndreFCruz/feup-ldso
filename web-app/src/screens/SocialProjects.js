import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { NotificationManager } from 'react-notifications';
import Route from 'react-router-dom/Route';
import { throttle } from 'throttle-debounce';
import { deepEqual } from '../utils';
import { setFaculty } from '../actions/facultyActions';
import { getProjects, getProjectDetails, addProject, editProject, deleteProject, searchProjects } from '../actions/projectsActions';
import Sidebar from '../components/Sidebar';
import GenericProject from '../components/GenericProject';

class SocialProjects extends Component {
    constructor(props) {
        super(props);

        this.state = {
            query: ''
        };
        this.updateQuery = this.updateQuery.bind(this);
    }

    componentDidMount() {
        this.updateFaculty();
    }

    componentDidUpdate(prevProps, prevState) {
        const { faculty, currProjEN, currProjPT, searchProjects, mapIds } = this.props;
        const { query } = this.state;

        if (prevProps.faculty !== faculty)
            this.updateFaculty();

        if (prevState.query !== query)
            searchProjects(faculty, query);

        if (Object.keys(prevProps.mapIds).length > Object.keys(mapIds).length)
            if (!global.__TEST__) NotificationManager.success('Successfully deleted project!');

        if (currProjEN != null && currProjPT != null) {
            if ((prevProps.currProjEN != null && prevProps.currProjEN.id === currProjEN.id && !deepEqual(prevProps.currProjEN, currProjEN)) ||
                (prevProps.currProjPT != null && prevProps.currProjPT.id === currProjPT.id && !deepEqual(prevProps.currProjPT, currProjPT)))
                if (!global.__TEST__) throttle(500, NotificationManager.success('Successfully edited project!'));
        }
    }

    updateFaculty() {
        const { match, faculty, getProjects, setFaculty } = this.props;

        setFaculty(match.params.faculty);
        getProjects(faculty, 'pt');
        getProjects(faculty, 'en');
    }

    updateQuery(query) {
        this.setState({
            query: query
        });
    }

    render() {
        const { loading, loadingAction, faculty, projectsEN, projectsPT, currProjEN, currProjPT, getProjectDetails, editProject, addProject, deleteProject } = this.props;

        return (
            <div style={contentStyle}>
                <Sidebar loading={loading} faculty={faculty} projectsEN={projectsEN} projectsPT={projectsPT} idProjEN={currProjEN != null ? currProjEN.id : null} action={getProjectDetails} search={this.updateQuery} query={this.state.query} />
                <Route exact path='/faculties/:faculty/projects' render={() =>
                    <GenericProject add={false} loading={loading} loadingAction={loadingAction} faculty={faculty} projEN={currProjEN} projPT={currProjPT} mainAction={editProject} delAction={deleteProject} />
                } />
                <Route exact path='/faculties/:faculty/projects/new' render={() =>
                    <GenericProject add={true} loading={loading} loadingAction={loadingAction} faculty={faculty} projEN={{}} projPT={{}} mainAction={addProject} delAction={null} />
                } />
            </div>
        );
    }
}

const contentStyle = {
    display: 'flex',
    flexDirecation: 'row'
};

SocialProjects.propTypes = {
    match: PropTypes.object.isRequired,
    faculty: PropTypes.string,
    loading: PropTypes.bool,
    loadingAction: PropTypes.bool,
    projectsEN: PropTypes.array,
    projectsPT: PropTypes.array,
    currProjEN: PropTypes.object,
    currProjPT: PropTypes.object,
    mapIds: PropTypes.object,
    setFaculty: PropTypes.func,
    getProjects: PropTypes.func,
    getProjectDetails: PropTypes.func,
    addProject: PropTypes.func,
    editProject: PropTypes.func,
    deleteProject: PropTypes.func,
    searchProjects: PropTypes.func
};

const mapStateToProps = ({ faculty, socialProjects }) => ({
    faculty: faculty.name,
    loading: socialProjects.loading,
    loadingAction: socialProjects.loadingAction,
    projectsEN: socialProjects.projectsEN,
    projectsPT: socialProjects.projectsPT,
    currProjEN: socialProjects.currProjEN,
    currProjPT: socialProjects.currProjPT,
    mapIds: socialProjects.idsMap
});

const mapDispatchToProps = {
    setFaculty,
    getProjects,
    getProjectDetails,
    addProject,
    editProject,
    deleteProject,
    searchProjects
};

export default connect(mapStateToProps, mapDispatchToProps)(SocialProjects);