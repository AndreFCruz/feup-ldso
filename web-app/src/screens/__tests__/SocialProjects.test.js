import React from 'react';
import SocialProjects from '../SocialProjects';
import Enzyme, { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

const middlewares = []; // you can mock any middlewares here if necessary
const mockStore = configureStore(middlewares);

function setup(projects) {

    const initialState = {
        faculty: {
            name: 'faup'
        },
        socialProjects: {
            loading: false,
            loadingAction: false,
            projectsEN: projects,
            projectsPT: projects,
            idsMap: {
                'idEn': 'idPt'
            },
            currProjEN: {
                'title': '[EN] Social project at FAUP',
                'short_description': 'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Etiam neque. Aliquam erat volutpat. Suspendisse sagittis ultrices augue. Pellentesque ipsum.',
                'content': '[EN] Aliquam sit amet varius lacus. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Sed tortor justo, ornare a vulputate in, mattis id mi. Donec varius, dui venenatis pharetra ultricies, ligula nulla ornare ex, quis mattis libero quam a tellus. Quisque quis est id nulla ornare semper at a nisi. Integer tincidunt egestas ornare. Nullam congue tellus arcu, eu ultricies mi pellentesque eget. Pellentesque massa sapien, suscipit sit amet lacus vel, faucibus faucibus elit. Cras in metus viverra, luctus magna ac, tristique leo. Aenean ultrices dui eget velit cursus, vitae suscipit purus varius. Aliquam erat volutpat. In porta metus et velit aliquet porta.',
                'start_date': '2017-01-01',
                'images': [
                    'https://picsum.photos/200/300/?random',
                    'https://picsum.photos/112/200/?random'
                ],
                'facultyId': 1
            },
            currProjPT: {
                'title': '[PT] Social project at FAUP',
                'short_description': 'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Etiam neque. Aliquam erat volutpat. Suspendisse sagittis ultrices augue. Pellentesque ipsum.',
                'content': '[EN] Aliquam sit amet varius lacus. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Sed tortor justo, ornare a vulputate in, mattis id mi. Donec varius, dui venenatis pharetra ultricies, ligula nulla ornare ex, quis mattis libero quam a tellus. Quisque quis est id nulla ornare semper at a nisi. Integer tincidunt egestas ornare. Nullam congue tellus arcu, eu ultricies mi pellentesque eget. Pellentesque massa sapien, suscipit sit amet lacus vel, faucibus faucibus elit. Cras in metus viverra, luctus magna ac, tristique leo. Aenean ultrices dui eget velit cursus, vitae suscipit purus varius. Aliquam erat volutpat. In porta metus et velit aliquet porta.',
                'start_date': '2017-01-01',
                'images': [
                    'https://picsum.photos/200/300/?random',
                    'https://picsum.photos/112/200/?random'
                ],
                'facultyId': 1
            }
        }
    };

    const props = {
        match: {
            params: {
                faculty: 'faup',
            }
        },
        getProjects: jest.fn(),
        getProjectDetails: jest.fn(),
        editProject: jest.fn(),
        addProject: jest.fn(),
        deleteProject: jest.fn(),
        setFaculty: jest.fn()
    };

    const wrapper = shallow(<SocialProjects {...props} />,
        {
            context: { store: mockStore(initialState) },
            disableLifecycleMethods: true
        });

    return wrapper;
}

describe('Social Projects Screen tests', () => {

    it('renders loading screen', () => {
        const projects = [];
        const wrapper = setup(projects);
        expect(wrapper.dive()).toMatchSnapshot();
    });
    
    it('calls updateQuery without crashing', () => {
        const projects = [];
        const wrapper = setup(projects);
        expect(wrapper.dive().instance().state.query).toBe('');
        wrapper.dive().instance().updateQuery('projectName');
        expect(wrapper.dive()).toMatchSnapshot();
    });  

    it('calls updateFaculty without crashing', () => {
        const projects = [];
        const wrapper = setup(projects);
        wrapper.dive().instance().updateFaculty();
        expect(wrapper.dive()).toMatchSnapshot();
    });  

    it('calls componentDidMount without crashing', () => {
        const projects = [];
        const wrapper = setup(projects);
        wrapper.dive().instance().componentDidMount();
        expect(wrapper.dive()).toMatchSnapshot();
    });

    it('calls componentDidUpdate without crashing', () => {
        const projects = [];

        const prevProps = {
            faculty: 'feup',
            currProjEN: {
                id: '69',
                content: 'mockup project'
            },
            mapIds: {}
        };

        const prevState = {
            query: 'mockup query'
        };

        const wrapper = setup(projects);
        wrapper.dive().instance().componentDidUpdate(prevProps, prevState);
        expect(wrapper.dive()).toMatchSnapshot();
    }); 
});