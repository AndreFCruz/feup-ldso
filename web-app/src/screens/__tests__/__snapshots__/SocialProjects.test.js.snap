// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot renders without crashing 1`] = `
<div>
  <div
    className="intro"
  >
    <img
      alt="application logo"
      src="test-file-stub"
      width="140"
    />
  </div>
  <div>
    <div
      className="navigator"
    >
      <h6
        className="fadeup"
      >
        <a
          aria-current={null}
          href="/faculties/fadeup"
          onClick={[Function]}
        >
          FADEUP
        </a>
      </h6>
      <h6
        className="faup"
      >
        <a
          aria-current={null}
          href="/faculties/faup"
          onClick={[Function]}
        >
          FAUP
        </a>
      </h6>
      <h6
        className="fbaup"
      >
        <a
          aria-current={null}
          href="/faculties/fbaup"
          onClick={[Function]}
        >
          FBAUP
        </a>
      </h6>
      <h6
        className="fcnaup"
      >
        <a
          aria-current={null}
          href="/faculties/fcnaup"
          onClick={[Function]}
        >
          FCNAUP
        </a>
      </h6>
      <h6
        className="fcup"
      >
        <a
          aria-current={null}
          href="/faculties/fcup"
          onClick={[Function]}
        >
          FCUP
        </a>
      </h6>
      <h6
        className="fdup"
      >
        <a
          aria-current={null}
          href="/faculties/fdup"
          onClick={[Function]}
        >
          FDUP
        </a>
      </h6>
      <h6
        className="fep"
      >
        <a
          aria-current={null}
          href="/faculties/fep"
          onClick={[Function]}
        >
          FEP
        </a>
      </h6>
      <h6
        className="feup"
      >
        <a
          aria-current={null}
          href="/faculties/feup"
          onClick={[Function]}
        >
          FEUP
        </a>
      </h6>
      <h6
        className="ffup"
      >
        <a
          aria-current={null}
          href="/faculties/ffup"
          onClick={[Function]}
        >
          FFUP
        </a>
      </h6>
      <h6
        className="flup"
      >
        <a
          aria-current={null}
          href="/faculties/flup"
          onClick={[Function]}
        >
          FLUP
        </a>
      </h6>
      <h6
        className="fmdup"
      >
        <a
          aria-current={null}
          href="/faculties/fmdup"
          onClick={[Function]}
        >
          FMDUP
        </a>
      </h6>
      <h6
        className="fmup"
      >
        <a
          aria-current={null}
          href="/faculties/fmup"
          onClick={[Function]}
        >
          FMUP
        </a>
      </h6>
      <h6
        className="fpceup"
      >
        <a
          aria-current={null}
          href="/faculties/fpceup"
          onClick={[Function]}
        >
          FPCEUP
        </a>
      </h6>
      <h6
        className="icbas"
      >
        <a
          aria-current={null}
          href="/faculties/icbas"
          onClick={[Function]}
        >
          ICBAS
        </a>
      </h6>
    </div>
    <div
      className="container null"
    >
      <div>
        <h1>
          Redirect to App login or another admin page
        </h1>
      </div>
    </div>
  </div>
  <div
    className="notification-container notification-container-empty"
  >
    <span />
  </div>
</div>
`;

exports[`Social Projects Screen tests renders loading screen 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SocialProjects
    addProject={[Function]}
    currProjEN={
      Object {
        "content": "[EN] Aliquam sit amet varius lacus. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Sed tortor justo, ornare a vulputate in, mattis id mi. Donec varius, dui venenatis pharetra ultricies, ligula nulla ornare ex, quis mattis libero quam a tellus. Quisque quis est id nulla ornare semper at a nisi. Integer tincidunt egestas ornare. Nullam congue tellus arcu, eu ultricies mi pellentesque eget. Pellentesque massa sapien, suscipit sit amet lacus vel, faucibus faucibus elit. Cras in metus viverra, luctus magna ac, tristique leo. Aenean ultrices dui eget velit cursus, vitae suscipit purus varius. Aliquam erat volutpat. In porta metus et velit aliquet porta.",
        "facultyId": 1,
        "images": Array [
          "https://picsum.photos/200/300/?random",
          "https://picsum.photos/112/200/?random",
        ],
        "short_description": "Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Etiam neque. Aliquam erat volutpat. Suspendisse sagittis ultrices augue. Pellentesque ipsum.",
        "start_date": "2017-01-01",
        "title": "[EN] Social project at FAUP",
      }
    }
    currProjPT={
      Object {
        "content": "[EN] Aliquam sit amet varius lacus. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Sed tortor justo, ornare a vulputate in, mattis id mi. Donec varius, dui venenatis pharetra ultricies, ligula nulla ornare ex, quis mattis libero quam a tellus. Quisque quis est id nulla ornare semper at a nisi. Integer tincidunt egestas ornare. Nullam congue tellus arcu, eu ultricies mi pellentesque eget. Pellentesque massa sapien, suscipit sit amet lacus vel, faucibus faucibus elit. Cras in metus viverra, luctus magna ac, tristique leo. Aenean ultrices dui eget velit cursus, vitae suscipit purus varius. Aliquam erat volutpat. In porta metus et velit aliquet porta.",
        "facultyId": 1,
        "images": Array [
          "https://picsum.photos/200/300/?random",
          "https://picsum.photos/112/200/?random",
        ],
        "short_description": "Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Etiam neque. Aliquam erat volutpat. Suspendisse sagittis ultrices augue. Pellentesque ipsum.",
        "start_date": "2017-01-01",
        "title": "[PT] Social project at FAUP",
      }
    }
    deleteProject={[Function]}
    editProject={[Function]}
    faculty="faup"
    getProjectDetails={[Function]}
    getProjects={[Function]}
    loading={false}
    loadingAction={false}
    match={
      Object {
        "params": Object {
          "faculty": "faup",
        },
      }
    }
    projectsEN={Array []}
    projectsPT={Array []}
    setFaculty={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Sidebar
          action={[Function]}
          faculty="faup"
          loading={false}
          projectsEN={Array []}
          projectsPT={Array []}
          query=""
          search={[Function]}
        />,
        <Route
          exact={true}
          path="/faculties/:faculty/projects"
          render={[Function]}
        />,
        <Route
          exact={true}
          path="/faculties/:faculty/projects/new"
          render={[Function]}
        />,
      ],
      "style": Object {
        "display": "flex",
        "flexDirecation": "row",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "action": [Function],
          "faculty": "faup",
          "idProjEN": undefined,
          "loading": false,
          "projectsEN": Array [],
          "projectsPT": Array [],
          "query": "",
          "search": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "exact": true,
          "path": "/faculties/:faculty/projects",
          "render": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "exact": true,
          "path": "/faculties/:faculty/projects/new",
          "render": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Sidebar
            action={[Function]}
            faculty="faup"
            loading={false}
            projectsEN={Array []}
            projectsPT={Array []}
            query=""
            search={[Function]}
          />,
          <Route
            exact={true}
            path="/faculties/:faculty/projects"
            render={[Function]}
          />,
          <Route
            exact={true}
            path="/faculties/:faculty/projects/new"
            render={[Function]}
          />,
        ],
        "style": Object {
          "display": "flex",
          "flexDirecation": "row",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "action": [Function],
            "faculty": "faup",
            "idProjEN": undefined,
            "loading": false,
            "projectsEN": Array [],
            "projectsPT": Array [],
            "query": "",
            "search": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "exact": true,
            "path": "/faculties/:faculty/projects",
            "render": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "exact": true,
            "path": "/faculties/:faculty/projects/new",
            "render": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "context": Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
